Feature/4116 add bundle state docs

leilenah opened the pull request:
### Reasons for making this change

- adds bundle state info to web interface so users can quickly investigate what each bundle state means

### Related issues

https://github.com/codalab/codalab-worksheets/issues/4116

### Screenshots

https://user-images.githubusercontent.com/25855750/177645489-9ed98d3b-15af-4d9d-8c80-3ed0fdeb1c7a.mov

### ready
![ready-closed](https://user-images.githubusercontent.com/25855750/177062545-4a98dcc4-908b-406f-bb1e-4ac03a935018.png)

### ready - tooltip open

![Screen Shot 2022-07-06 at 2 05 36 PM](https://user-images.githubusercontent.com/25855750/177645543-3f4f8ec0-b553-496f-bcbb-12a13956e73c.png)

### staged
![staged](https://user-images.githubusercontent.com/25855750/177062577-5edb728b-5f3a-4995-b5ee-c72fe40f287a.png)

### bundle view
![bundle-view](https://user-images.githubusercontent.com/25855750/177062581-d828a5b9-7f73-4c53-abf5-ba4b22fd7a47.png)


### Checklist

* [x] I've added a screenshot of the changes, if this is a frontend change
* [ ] I've added and/or updated tests, if this is a backend change
* [x] I've run the [pre-commit.sh](https://github.com/codalab/codalab-worksheets/blob/master/pre-commit.sh) script
* [x] I've updated [docs](https://github.com/codalab/codalab-worksheets/tree/master/docs), if needed

Commit:
Add bundle state docs.

Commit:
Add StateTooltip

Commit:
Add state tooltip to bundle view

Commit:
Update StateTooltip constants

leilenah commented:
@epicfaace comments resolved ✅ 
cc @pranavjain

pranavjain commented:
@leilenah Thanks for the change.

The current UI (as per the video) doesn't help the user to understand the details/flow of the states. I think you should use numbers instead of the list so that it shows a flow.

Also, please add the same icon on the sidebars where we view the bundle details. We can show a single-line description of the state below the state box in the sidebar. A more detailed version can always be presented via docs.


<img width="427" alt="image" src="https://user-images.githubusercontent.com/10134824/175177686-61cbdb85-b6fe-4f65-8d3e-69d973d42d62.png">

Commit:
update state tooltip UI and add to sidebar

leilenah commented:
@pranavjain Sidebar & tooltip content updated (see screenshots). 

> We can show a single-line description of the state below the state box in the sidebar.

By ☝️  you mean a dynamic state description that changes as the state changes? Given the presence of the tooltip, I feel that this would be a bit excessive / redundant. However if you feel strongly, I can add this in. Lmk your thoughts.

pranavjain commented:
> @pranavjain Sidebar & tooltip content updated (see screenshots).
> 
> > We can show a single-line description of the state below the state box in the sidebar.
> 
> By ☝️ you mean a dynamic state description that changes as the state changes? Given the presence of the tooltip, I feel that this would be a bit excessive / redundant. However if you feel strongly, I can add this in. Lmk your thoughts.

Yes, don't add it as a part of the tooltip but as a separate row below the state (or after the "ready" status).

It would be easy to take a call once we have some design screenshots. Thanks for the amazing work!

leilenah committed:
Add bundle state details to sidebar.

leilenah commented:
@pranavjain changes pushed 🙃 

### Example 1 - staged
![Screen Shot 2022-06-23 at 8 41 01 PM](https://user-images.githubusercontent.com/25855750/175457755-e65004e5-19ec-496b-8f5e-0868020ed885.png)

### Example 2 - ready
![Screen Shot 2022-06-23 at 8 41 15 PM](https://user-images.githubusercontent.com/25855750/175457776-89ce9f1e-832e-4282-bdf0-79e6a4971d83.png)

### Example 3 - killed 
![Screen Shot 2022-06-23 at 8 41 31 PM](https://user-images.githubusercontent.com/25855750/175457794-25ad2a47-474c-41d5-a9ce-eddc8702507b.png)

pranavjain commented:
I like this. Thanks!

Can you please also make sure that the same change is propagated on the full bundle details page? Also, how would the users be redirected to documentation with details about each state?

leilenah commented:
> I like this. Thanks!
> 
> Can you please also make sure that the same change is propagated on the full bundle details page? Also, how would the users be redirected to documentation with details about each state?

@pranavjain can do 👍 and if a user clicks the tooltip icon, the full documentation will open in a new page.

leilenah committed:
Update state tooltip

leilenah commented:
@percyliang @pranavjain all comments are addressed in the latest commit. Here's a working doc for the bundle state descriptions if you want to make more changes:

https://docs.google.com/document/d/1mR6DZAaHt_x7T_H8tA96F7znQLYcwnrvfaVWwQJTH9Q/edit#

leilenah committed:
Merge branch 'master' into feature/4116-add-bundle-state-docs

leilenah committed:
Update bundle state docs and refactor tooltip.

leilenah commented:
@percyliang @pranavjain @epicfaace this is ready for re-review. Here are some updated UI screenshots:

### Detail Sidebar
![detail-sidebar](https://user-images.githubusercontent.com/25855750/176595433-f1d63737-85d0-4932-aa83-fc37f71a3d02.png)

### Run Bundle - Tooltip Hover
![run-bundle](https://user-images.githubusercontent.com/25855750/176595474-ae1527fb-e00b-4b6b-acea-6381cfd515ee.png)

### Uploaded Bundle - Tooltip Hover
![uploaded-bundle](https://user-images.githubusercontent.com/25855750/176595528-0f8ba4e7-a465-4b90-8605-1a9da77dacc5.png)

### Make Bundle - Tooltip Hover
![make-bundle](https://user-images.githubusercontent.com/25855750/176595555-4e798d0f-ff66-481c-a3cb-b4b81b835356.png)

### Bundle View
![bundle-view](https://user-images.githubusercontent.com/25855750/176595585-f3db3bcb-d182-4c4f-868a-4543e93755d3.png)

pranavjain commented:
Thanks for sharing the screenshots.

In the case of the sidebar, I personally liked the `Status: Staged` written on a single line and then the description on the line below that. But, it is not a big deal.

@percyliang Your call.

leilenah commented:
> In the case of the sidebar, I personally liked the `Status: Staged` written on a single line and then the description on the line below that.

@pranavjain I prefer that also. Here's another option:

![Screen Shot 2022-06-29 at 10 26 52 PM](https://user-images.githubusercontent.com/25855750/176599655-0bb7a31d-7f1b-4b83-bd4b-5e97ecde51c1.png)

Basically what I had before, but without "**State Details:**" preceding the detail text. (Also it would be the updated icon, not the one shown here). cc @percyliang

pranavjain commented:
@leilenah How about we create a borderless table kind of format like Wikipedia does? `command` option might be an exception here though. This will make the overall design much cleaner.

Thoughts? Would it be too much work/change?

![image](https://user-images.githubusercontent.com/10134824/176602199-458e9ef1-d594-4e7e-bf41-0226cc4852d1.png)

leilenah commented:
> @leilenah How about we create a borderless table kind of format like Wikipedia does? `command` option might be an exception here though. This will make the overall design much cleaner.
> 
> Thoughts? Would it be too much work/change?
> 
> ![image](https://user-images.githubusercontent.com/10134824/176602199-458e9ef1-d594-4e7e-bf41-0226cc4852d1.png)

Per Slack convo, I will create a separate issue for this.

leilenah commented:
looking into the failing end-to-end ui tests

leilenah committed:
Bundle state detail refactor [part 1]

leilenah committed:
bundle state detail refactor [part 2]

leilenah commented:
@percyliang @pranavjain I have implemented the changes that we discussed last week. Updated screenshots have been added to the description. Plz review.

#### Updates:
1. The tooltip uses downward arrows now
2. I have added a white border to indicate current state (bolding the current state wasn't working out 😅)
3. Colored state text block has been moved back up next to `State` heading
4. Bundle states are now returned by the rest server instead of hardcoded at the component level
5. Run status will be shown for `running` bundles
6. Staged status will be shown for `staged` bundles (no one asked for this, but it seemed appropriate)

leilenah committed:
round out bundle state tooltip current state border

leilenah committed:
Tweak bundle state docs.

leilenah committed:
Merge branch 'master' into feature/4116-add-bundle-state-docs

leilenah committed:
Merge branch 'master' into feature/4116-add-bundle-state-docs

leilenah committed:
Update bundle state docs.

leilenah committed:
Update Bundle-Lifecycle.md

percyliang commented:
Graphics look good. Could you make it "worker_offline" to be more precise?

leilenah commented:
> Graphics look good. Could you make it "worker_offline" to be more precise?

@percyliang certainly. Had to bump the font for that state down a smidge tho.

![run-bundle-states](https://user-images.githubusercontent.com/25855750/178583759-faf7dffd-bb67-4034-8bbb-cff86e3b6935.png)

leilenah committed:
Refine bundle state docs.

leilenah commented:
Issue to debug / improve `staged_status`: https://github.com/codalab/codalab-worksheets/issues/4139

cc @percyliang

leilenah committed:
Add error logging to get_bundle_state_details

leilenah committed:
Remove defensive checks and logging from get_bundle_state_details.

percyliang reviewed (approved):
Great!  Could you also add the original files for the figures to the repo?

leilenah commented:
@percyliang I made those in Lucidchart [here](https://lucid.app/lucidchart/ab715d01-9722-4731-96d1-638f553fde64/edit?viewport_loc=-33%2C-36%2C2025%2C1613%2C0_0&invitationId=inv_de3f90df-7870-421b-87db-4f14479ab009#) so there's not really a way for me to add editable versions to the repo.

But I have added that Lucidchart link to the shared CodaLab folder [here](https://docs.google.com/document/d/1mR6DZAaHt_x7T_H8tA96F7znQLYcwnrvfaVWwQJTH9Q/edit#) in case someone needs to modify the figures in the future.

leilenah committed:
Merge branch 'master' into feature/4116-add-bundle-state-docs

epicfaace committed:
Merge branch 'master' into feature/4116-add-bundle-state-docs

leilenah committed:
Bundle state_details test update

leilenah committed:
Merge branch 'feature/4116-add-bundle-state-docs' of https://github.com/codalab/codalab-worksheets into feature/4116-add-bundle-state-docs

leilenah committed:
Fix GetBundleStateDetailsTest

leilenah committed:
Fix UtilTest

leilenah committed:
Fix competition unit test
